test/e2e/bandwidth_test.go

```diff
@@ -98,3 +100,73 @@ func TestBenchmarkBandwidthServiceRemoteAccess(t *testing.T) {
 	skipIfNumNodesLessThan(t, 2)
 	benchmarkBandwidthService(t, masterNodeName(), workerNodeName(1))
 }
+
+func TestPodTrafficShaping(t *testing.T) {
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+	tests := []struct {
+		name string
+		// The bandwidths' unit is Mbits/sec.
+		clientEgressBandwidth  int
+		serverIngressBandwidth int
+		// Use a range to allow a certain deviation.
+		expectedMinBandwidth int
+		expectedMaxBandwidth int
+	}{
+		{
+			name:                   "limited by egress bandwidth",
+			clientEgressBandwidth:  100,
+			serverIngressBandwidth: 200,
+			expectedMinBandwidth:   90,
+			expectedMaxBandwidth:   110,
+		},
+		{
+			name:                   "limited by ingress bandwidth",
+			clientEgressBandwidth:  300,
+			serverIngressBandwidth: 200,
+			expectedMinBandwidth:   180,
+			expectedMaxBandwidth:   220,
+		},
+	}
+	for i, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			clientPodName := fmt.Sprintf("client-a-%d", i)
+			serverPodName := fmt.Sprintf("server-a-%d", i)
+			if err := data.createPodOnNode(clientPodName, masterNodeName(), perftoolImage, nil, nil, nil, nil, false, func(pod *v1.Pod) {
+				pod.Annotations = map[string]string{
+					"kubernetes.io/egress-bandwidth": fmt.Sprintf("%dM", tt.clientEgressBandwidth),
+				}
+			}); err != nil {
+				t.Fatalf("Error when creating the perftest client Pod: %v", err)
+			}
+			defer deletePodWrapper(t, data, clientPodName)
+			if err := data.podWaitForRunning(defaultTimeout, clientPodName, testNamespace); err != nil {
+				t.Fatalf("Error when waiting for the perftest client Pod: %v", err)
+			}
+			if err := data.createPodOnNode(serverPodName, masterNodeName(), perftoolImage, nil, nil, nil, []v1.ContainerPort{{Protocol: v1.ProtocolTCP, ContainerPort: iperfPort}}, false, func(pod *v1.Pod) {
+				pod.Annotations = map[string]string{
+					"kubernetes.io/ingress-bandwidth": fmt.Sprintf("%dM", tt.serverIngressBandwidth),
+				}
+			}); err != nil {
+				t.Fatalf("Error when creating the perftest server Pod: %v", err)
+			}
+			defer deletePodWrapper(t, data, serverPodName)
+			podBIP, err := data.podWaitForIP(defaultTimeout, serverPodName, testNamespace)
+			if err != nil {
+				t.Fatalf("Error when getting the perftest server Pod's IP: %v", err)
+			}
+			stdout, _, err := data.runCommandFromPod(testNamespace, clientPodName, "perftool", []string{"bash", "-c", fmt.Sprintf("iperf3 -c %s -f m -O 1|grep sender|awk '{print $7}'", podBIP)})
+			if err != nil {
+				t.Fatalf("Error when running iperf3 client: %v", err)
+			}
+			stdout = strings.TrimSpace(stdout)
+			actualBandwidth, _ := strconv.ParseFloat(strings.TrimSpace(stdout), 64)
+			t.Logf("Actual bandwidth: %v Mbits/sec", actualBandwidth)
+			assert.Greater(t, actualBandwidth, float64(tt.expectedMinBandwidth))
+			assert.Less(t, actualBandwidth, float64(tt.expectedMaxBandwidth))
```

Comment:
the `assert` package has the `InEpsilon` assertion that you may want to use here

https://godoc.org/github.com/stretchr/testify/assert#InEpsilon

Reply:
Thanks! this is exactly what I'm looking for.